File: firebase/functions/src/normalize.ts
import {createHash} from "crypto";

/**
 * Normalize sender for idempotency: trim + lowercase.
 */
export function normalizeSender(sender: string): string {
  return sender.trim().toLowerCase();
}

/**
 * Collapse internal whitespace in SMS body to a single space.
 */
export function normalizeBody(body: string): string {
  return body.trim().replace(/\s+/g, " ");
}

/**
 * Floor an ISO8601 timestamp to the start of its minute (UTC).
 * Returns ISO8601 string used in the idempotency key.
 */
export function floorToMinute(receivedAt: string): string {
  const date = new Date(receivedAt);
  if (Number.isNaN(date.getTime())) {
    throw new Error("Invalid receivedAt: must be ISO8601");
  }
  date.setUTCSeconds(0, 0);
  return date.toISOString();
}

/**
 * idempotencyKey = sha256(senderNorm + "|" + bodyNorm + "|" + minuteBucket)
 */
export function computeIdempotencyKey(
  sender: string,
  body: string,
  receivedAt: string,
): string {
  const senderNorm = normalizeSender(sender);
  const bodyNorm = normalizeBody(body);
  const minuteBucket = floorToMinute(receivedAt);
  const payload = `${senderNorm}|${bodyNorm}|${minuteBucket}`;
  return createHash("sha256").update(payload, "utf8").digest("hex");
}

/**
 * Hash a bearer token for comparison against device_tokens.tokenHash.
 */
export function hashToken(token: string): string {
  return createHash("sha256").update(token, "utf8").digest("hex");
}

File: lib/core/utils/normalize.dart
import 'package:crypto/crypto.dart';
import 'dart:convert';

/// Normalizes sender for idempotency: trim + lowercase.
String normalizeSender(String sender) => sender.trim().toLowerCase();

/// Collapses internal whitespace in SMS body for idempotency.
String normalizeBody(String body) {
  return body.trim().replaceAll(RegExp(r'\s+'), ' ');
}

/// Floors [receivedAt] to the start of its minute (UTC).
DateTime floorToMinute(DateTime receivedAt) {
  return DateTime.utc(
    receivedAt.year,
    receivedAt.month,
    receivedAt.day,
    receivedAt.hour,
    receivedAt.minute,
  );
}

/// Client-side preview of the server idempotency key (sha256 hex).
///
/// Matches Cloud Function contract: sha256(senderNorm|bodyNorm|minuteBucket).
String computeIdempotencyKey({
  required String sender,
  required String body,
  required DateTime receivedAt,
}) {
  final senderNorm = normalizeSender(sender);
  final bodyNorm = normalizeBody(body);
  final minuteBucket = floorToMinute(receivedAt.toUtc()).toIso8601String();
  final payload = '$senderNorm|$bodyNorm|$minuteBucket';
  return sha256.convert(utf8.encode(payload)).toString();
}
